src/systemctl/systemctl-show.c

```diff
@@ -2103,26 +2104,85 @@ static int show_one(
 static int get_unit_dbus_path_by_pid(
```

Comment:
Wouldn't it make more sense to put your partially pidfd-oriented implementation here in get_unit_dbus_path_by_pid()?  After all, the C API is still just taking a pid, not a pidfd argument... it seems like more of an implementation detail wether the underlying function acquires the pidfd or not.

Maybe it's just a necessarily confused transitional state of things, but it seems inappropriate to have `get_unit_dbus_path_by_pidfd()` exist at all when its API is still taking a numeric pid argument.

I'd rather just use the pidfd feature to close up the race a bit as you've done but in `get_unit_dbus_path_by_pid()`.  At least its API is honest by receiving a PID which will *always* have a race vs. recycling.  Then make `get_unit_dbus_path_by_pidfd()` accept a pidfd, just having no callers until the rest of the code can pivot to using pidfds and start calling the by_pidfd variant.

Do we need to keep a 'GetUnitByPID' method for backwards compat?  Because it's arguably just buggy to ever do by pid...

Reply:
we need to retain compat for old kernels for now (our baseline is 3.15 for now, pidfd_open is 5.3). Which sucks, but it is how it is.

In general I am a big fan of making the "clean" codepaths the ones we want this to work like, and then adding compat codepaths for the "fallback" codepaths. i.e. always think in pidfd everywhere use that as primary conduit for everything, and use pid fallbacks only whenver we have to.

Reply:
This is not really an API, it's a local static helper. The user input has to be a PID, given it's a console command, you can't really (well, easily) pass an FD there. There's a second helper because we have to take the fallback path in two cases (no syscall or no dbus method).